nni/mutable/mutable.py

```diff
@@ -0,0 +1,482 @@
+
+    def robust_default(self, memo: Sample | None = None, retries: int = 1000) -> Any:
+        """Return the default value of the mutable.
+        Will retry with :meth:`random` in case of failure.
+
+        It's equivalent to the following pseudo-code::
+
+            for attempt in range(retries + 1):
+                try:
+                    if attempt == 0:
+                        return self.default()
+                    else:
+                        return self.random()
+                except SampleValidationError:
+                    pass
+
+        Parameters
+        ----------
+        memo
+            A dict of mutable labels and their default values.
+            Use this to share the sampled value among mutables with the same label.
+        retries
+            If the default sample is not valid, we will retry to invoke
+            :meth:`random` for ``retries`` times, until a valid sample is found.
+            Otherwise, an exception will be raised, complaining that no valid sample is found.
+        """
+        sample: Sample = {} if memo is None else memo
+        for attempt in range(retries + 1):
+            try:
+                sample_copy = copy.copy(sample)
+                if attempt == 0:
+                    rv = self.default(sample_copy)
+                else:
+                    rv = self.random(sample_copy)
+                sample.update(sample_copy)
+                return rv
+            except SampleValidationError:
+                if attempt == retries:
+                    raise ValueError(
+                        f'Cannot find a valid default sample after {retries} retries, for {self}. '
+                        'Please either set `default_value` manually, or loosen the constraints.')
+
+        raise RuntimeError('This should not happen.')
+
+    def random(self, memo: Sample | None = None, random_state: RandomState | None = None) -> Any:
+        """Randomly sample a value of the mutable. Used in random strategy.
+        The returned value should be one of the possible results of :meth:`freeze`.
+
+        The default implementation of :meth:`random` is to call :meth:`random`
+        on each of the simplified values and then freeze the result.
+
+        It's possible that :meth:`random` raises :exc:`~nni.mutable.exception.SampleValidationError`,
+        e.g., in cases when constraints are violated.
+
+        Parameters
+        ----------
+        memo
+            A dict of mutable labels and their random values.
+            Use this to share the sampled value among mutables with the same label.
+        """
+        sample: Sample = {} if memo is None else memo
+        if random_state is None:
+            random_state = RandomState()
+        for mutable in self.simplify().values():
+            mutable.random(sample, random_state)
+        return self.freeze(sample)
+
+    def grid(self, memo: Sample | None = None, granularity: int | None = None) -> Iterable[Any]:
```

Comment:
So the caller of this function should iterate granularity in range(inf), and dedup with a set?

Reply:
That's correct.

But I imagine most use cases of `grid` should be with a certain granularity level.